### ISSUE TITLE
Deep learning for... chess

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11096]
Deep learning for... chess<br>
<br>
https://ift.tt/a4qLGQW<br>
<br>
Erik Bernhardsson<br>
<br>

<div><div><div><div><article><h1>Deep learning for... chess</h1>
<time datetime="2014-11-29T00:00:00Z">2014-11-29</time><p>I've been meaning to learn <a href="http://deeplearning.net/software/theano/">Theano</a> for a while and I've also wanted to build a chess AI at some point. So why not combine the two? That's what I thought, and I ended up spending way too much time on it. I actually built most of this back in September but not until Thanksgiving did I have the time to write a blog post about it.</p>
<p><strong>What's the theory?</strong></p>
<p>Chess is a game with a finite number of states, meaning if you had infinite computing capacity, you could actually <a href="http://en.wikipedia.org/wiki/Solving_chess">solve chess</a>. Every position in chess is either a win for white, a win for black, or a forced draw for both players. We can denote this by the function $$ f(\mbox{position}) $$ . If we had an infinitely fast machine we could compute this by</p>
<ol>
<li>Assign all the final positions the value $$ {-1, 0, 1} $$ depending on who wins.</li>
<li>Use the recursive rule</li>
</ol>
<p>$$ f(p) = \max_{p \rightarrow p’} -f(p’) $$</p>
<p>where $$ p \rightarrow p’ $$ denotes all the legal moves from position $$ p $$ . The minus sign is because the players alternate between positions, so if position $$ p $$ is white's turn, then position $$ p’ $$ is black turns (and vice versa). This is the same thing as <a href="http://en.wikipedia.org/wiki/Minimax">minimax</a>.</p>
<p>There's approximately <a href="http://en.wikipedia.org/wiki/Shannon_number">$$ 10^{43} $$ positions</a>, so there's no way we can compute this. We need to resort to approximations to $$ f(p) $$ .</p>
<p><strong>What's the point of using machine learning for this?</strong></p>
<p>What machine learning really boils down to is approximating functions given data. So assuming we can get a lot of data to learn this from, we can learn this function $$ f(p) $$ . Once we have a model, an objective, and training data, we can go knock ourselves out.</p>
<p>I downloaded 100M games from <a href="http://www.ficsgames.org/download.html">FICS Games Database</a> and began training a machine learning model. My function $$ f(p) $$ is learned from data by using two principles</p>
<ol>
<li>Players will choose an optimal or near-optimal move. This means that for two position in succession $$ p \rightarrow q $$ observed in the game, we will have $$ f(p) = -f(q) $$ .</li>
<li>For the same reason above, going from $$ p $$ , not to $$ q $$ , but to a <em>random</em> position $$ p \rightarrow r $$ , we must have $$ f(r) &gt; f(q) $$ because the random position is better for the next player and worse for the player that made the move.</li>
</ol>
<p><strong>The model</strong></p>
<p>We construct $$ f(p) $$ as a 3 layer deep 2048 units wide artificial neural network, with rectified linear units in each layer. The input is a 8 * 8 * 12 = 768 wide layer which indicates whether each piece (there are 12 types) is present in each square (there are 8 * 8 squares). After three matrix multiplications (each followed by a nonlinearity), there's a final dot product with a 2048-wide vector to condense it down to a single value.</p>
<p><img alt="image" src="https://erikbern.com/assets/2014/11/chess-architecture-1.png"></p>
<p>In total there's roughly 10M unknown parameters in the network.</p>
<p>To train the network, I present it with $$ (p, q, r) $$ triplets. I feed it through the network. Denoting by $$ S(x) = 1 / (1 + exp(-x)) $$ , the sigmoid function, the total objective is:</p>
<p>$$ sum_{(p, q, r)} \log S(f(q) - f(r)) + \kappa \log (f(p) + f(q)) + \kappa \log (-f(q) - f(p)) $$</p>
<p> </p>
<p>This is the log likelihood of the “soft” inequalities $$ f(r) &gt; f(q) $$ , $$ f(p) &gt; -f(q) $$ , and $$ f(p) &lt; -f(q) $$ . The last two are just a way of expressing a “soft” equality $$ f(p) = -f(q) $$ . I also use $$ \kappa $$ to put more emphasis on getting the equality right. I set it to 10.0. I don't think the solution is super sensitive to the value of $$ \kappa $$ .</p>
<p><img alt="image" src="https://erikbern.com/assets/2014/11/chess-architecture.png"></p>
<p>Notice that the function we learn <em>has no idea about the rules of chess.</em> We're not even teaching it how each piece move. We make sure the model has the expressiveness to work out legal moves, but we don't encode any information about the game itself. The model learns this information by observing tons of chess games.</p>
<p>Note that I'm also not trying to learn anything from <em>who won the game.</em> The reason is that the training data is full of games played by amateurs. If a grandmaster came into the middle of a game, s/he could probably completely turn it around. This means the final score is a pretty weak label. Still, even an amateur player probably makes near-optimal moves for most time.</p>
<p><strong>Training the model</strong></p>
<p>I rented a GPU instance from AWS and trained it on 100M games for about four days using stochastic gradient descent with Nesterov momentum. I put all (p, q, r) triplets into a <a href="http://www.h5py.org/">HDF5 data file</a>. I was messing around with learning rates for a while but after a while I realized I just wanted something that would give me good results in a few days. So I ended using a slightly unorthodox learning rate scheme: $$ 0.03 \cdot \exp(-\mbox{time in days}) $$ . Since I had so much training data, regularization wasn't necessary, so I wasn't using either dropout or L2 regularization.</p>
<p>A trick I did was to encode the boards as 64 bytes and then transform the board into a 768 units wide float vector on the GPU. This gave a pretty substantial performance boost since there's a lot less I/O.</p>
<p><strong>How does a chess AI work?</strong></p>
<p>Every chess AI starts with some function $$ f(p) $$ that approximates the value of the position. This is known as <a href="http://en.wikipedia.org/wiki/Evaluation_function">evaluation function</a>.</p>
<p>This function is also combined with a deep search of many millions of positions down the game tree. It turns out that an approximation of $$ f(p) $$ is just a small part of the playing chess well. All chess AI's focus on smart search algorithms, but the number of positions explode exponentially down the search tree, so in practice you can't go deeper than say 5-10 positions ahead. What you do is you use some approximation to evaluate leaf nodes and then use some variety of <a href="http://en.wikipedia.org/wiki/Negamax">negamax</a> to evaluate a game tree of a bunch of possible next moves.</p>
<p>By applying some smart searching algorithm, we can take pretty much any approximation and make it better. Chess AI's typically start with some simple evaluation function like: every pawn is worth 1 point, every knight is worth 3 points, etc.</p>
<p>We're going to take the function we learned and use it to evaluate leaves in the game tree. Then try to search deep. So we're first going to learn the function $$ f(p) $$ from data, then we're going to plug it into a search algorithm.</p>
<p><strong>Does it work?</strong></p>
<p>I coined my chess engine <em>Deep Pink</em> as an homage to <a href="http://en.wikipedia.org/wiki/Deep_Blue_%28chess_computer%29">Deep Blue</a>. As it turns out, the function we learn can definitely play chess. It beats me, every time. But I'm a horrible chess player.</p>
<p>Does Deep Pink beat existing chess AI's? <strong>Sometimes</strong></p>
<p>I pit it against another chess engine: <a href="https://github.com/thomasahle/sunfish">Sunfish</a> by Thomas Dybdahl Ahle. Sunfish is written entirely in Python. The reason I chose to stick to the same language was that I didn't want this to be an endless exercise of making fast move generation. Deep Pink also relies heavily on quick move generation, and I didn't want to spend weeks working out edge cases with bitmaps in C++ to be able to compete with the state of the art engines. That would just be an arms race. So to be able to establish something useful, I picked a pure Python engine.</p>
<p>The obvious thing in hindsight is: the main thing you want out of any evaluation function $$ f(p) $$ isn't accuracy, it's <strong>accuracy per time unit</strong>. It doesn't matter that one evaluation function is slightly better than another if it's ten times slower, because you can take the fast (but slightly worse) evaluation function and search more nodes in the game tree. So you really want to take into account the time spent by the engine. Without further ado, here's some results of playing against the engine many times:</p>
<p><img alt="image" src="https://erikbern.com/assets/2014/11/scatter-plot1.png"></p>
<p>Notice the log-scale. The x-axis and y-axis aren't super relevant here, the main thing is the distance to the diagonal, because that tells us which engine spent more CPU time. Every game I randomized the parameters for each engine: the max depth for Deep Pink, and the max number of nodes for Sunfish. (I didn't include draws because both engines struggle with it).</p>
<p>Not surprisingly, the more time advantage either side has, the better it plays. <strong>Overall, Sunfish is better, winning the majority of the games, but Deep Pink probably wins 1/3 of the time.</strong> I'm actually pretty encouraged by this. I think with some optimizations, Deep Pink could actually play substantially better:</p>
<ul>
<li>Better search algorithm. I'm currently using <a href="http://en.wikipedia.org/wiki/Negamax#NegaMax_with_Alpha_Beta_Pruning">Negamax with alpha-beta pruning</a>, whereas Sunfish uses <a href="http://en.wikipedia.org/wiki/MTD-f">MTD-f</a>
</li>
<li>Better evaluation function. Deep Pink plays pretty aggressively, but makes a lot of dumb mistakes. By generating “harder” training examples (ideally fed from mistakes it made) it should learn a better model</li>
<li>Faster evaluation function: It might be possible to train a smaller (but maybe deeper) version of the same neural network</li>
<li>Faster evaluation function: I didn't use the GPU for playing, only for training.</li>
</ul>
<p>Obviously the real goal wouldn't be to beat Sunfish, but one of the “real” chess engines out there. But for that, I would have to write carefully tuned C++ code, and I'm not sure it's the best way to spend my time.</p>
<p><strong>Summary</strong></p>
<p>I'm encouraged by this. I think it's really cool that</p>
<ol>
<li>It's possible to learn an evaluation function directly from raw data, with no preprocessing</li>
<li>A fairly slow evaluation function (several orders of magnitude slower) can still play well if it's more accurate</li>
</ol>
<div><p>
I'm pretty curious to see if this could fare well for <a href="http://en.wikipedia.org/wiki/Computer_Go">Go</a> or other games where AI's still don't play well. Either way, the conclusions above come with a million caveats. The biggest one is obviously that I haven't challenged a “real” chess engine. I'm not sure if I have the time to start hacking on chess engines, but if anyone is interested, I've <a href="https://github.com/erikbern/deep-pink">put all the source code up on Github</a>.
</p></div>
<p> </p>
<strong>
Tagged with: <a href="https://erikbern.com/tags/math">math</a>
</strong>
</article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Erik Bernhardsson <br>
<br>
April 16, 2024 at 10:43AM
